### PR TITLE
mosfire sequence bug

### PR DESCRIPTION
**Pull request by debora-pe**
This is to address issue #1252 

It seemed to be only a problem of masking. Slits without bspline fits were process by `fit2illumflat` instead of being masked.

Tested on the data of the PypeIt user that raised the issue.

Also Dev Suite passed:
```
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-08-24T18:48:49.889338
Testing Completed at 2021-08-25T08:41:42.263199
Total Time: 13:52:52.373861
```

And unit tests too.
```
pytest . -W ignore
====================================================== test session starts ======================================================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 287 items                                                                                                             

test_alignments.py .                                                                                                      [  0%]
test_arc.py .                                                                                                             [  0%]
test_archive.py ..                                                                                                        [  1%]
test_arcimage.py ..                                                                                                       [  2%]
test_biasframe.py ..                                                                                                      [  2%]
test_bitmask.py ....                                                                                                      [  4%]
test_bpmimage.py ..                                                                                                       [  4%]
test_bspline.py ..........                                                                                                [  8%]
test_buildimage.py .                                                                                                      [  8%]
test_calibrations.py .......                                                                                              [ 11%]
test_chk_calibs.py ..                                                                                                     [ 11%]
test_coadd.py .                                                                                                           [ 12%]
test_collate_1d.py ............                                                                                           [ 16%]
test_cooked.py .                                                                                                          [ 16%]
test_coremeta.py .                                                                                                        [ 17%]
test_datacontainer.py ........                                                                                            [ 19%]
test_detector.py ...                                                                                                      [ 20%]
test_display.py .                                                                                                         [ 21%]
test_edgetrace.py .                                                                                                       [ 21%]
test_fitting.py ....                                                                                                      [ 22%]
test_flatfield.py .                                                                                                       [ 23%]
test_flexure.py ..                                                                                                        [ 24%]
test_flux.py ....                                                                                                         [ 25%]
test_fluxspec.py .....                                                                                                    [ 27%]
test_frametype.py .                                                                                                       [ 27%]
test_history.py ....                                                                                                      [ 28%]
test_load.py .                                                                                                            [ 29%]
test_load_images.py ................                                                                                      [ 34%]
test_maskdesign_matching.py ...                                                                                           [ 35%]
test_masterframe.py .                                                                                                     [ 36%]
test_match.py ....                                                                                                        [ 37%]
test_metadata.py ......                                                                                                   [ 39%]
test_moment.py ......                                                                                                     [ 41%]
test_msgs.py ..                                                                                                           [ 42%]
test_opticalmodel.py ....                                                                                                 [ 43%]
test_parse.py ...                                                                                                         [ 44%]
test_pca.py ........                                                                                                      [ 47%]
test_processimage.py ..                                                                                                   [ 48%]
test_procimg.py .                                                                                                         [ 48%]
test_pydl.py .                                                                                                            [ 49%]
test_pypeit.py ..                                                                                                         [ 49%]
test_pypeitimage.py .                                                                                                     [ 50%]
test_pypeitpar.py ................................                                                                        [ 61%]
test_pypeitsetup.py .......                                                                                               [ 63%]
test_qa.py .                                                                                                              [ 64%]
test_runpypeit.py ..                                                                                                      [ 64%]
test_scienceimage.py ...                                                                                                  [ 65%]
test_scripts.py ..............                                                                                            [ 70%]
test_setups.py .....................                                                                                      [ 78%]
test_skysub.py ..                                                                                                         [ 78%]
test_slitmask.py ...                                                                                                      [ 79%]
test_slittrace.py ....                                                                                                    [ 81%]
test_spec2dobj.py ......                                                                                                  [ 83%]
test_specobj.py .......                                                                                                   [ 85%]
test_specobjs.py .....                                                                                                    [ 87%]
test_spectrographs.py ....................                                                                                [ 94%]
test_syncspec.py .                                                                                                        [ 94%]
test_telluric.py ..                                                                                                       [ 95%]
test_traceimage.py .                                                                                                      [ 95%]
test_utils.py ...                                                                                                         [ 96%]
test_vcorr.py ..                                                                                                          [ 97%]
test_wavemodel.py .                                                                                                       [ 97%]
test_wavetilts.py ....                                                                                                    [ 99%]
test_wvcalib.py ..                                                                                                        [100%]

=============================================== 287 passed in 2372.32s (0:39:32) ================================================
```

### CODE REVIEW

File: pypeit/flatfield.py [hunk 1]
```diff
@@ -265,7 +265,7 @@ def fit2illumflat(self, slits, frametype='illum', initial=False, flexure_shift=N
         # Loop
         for slit_idx in range(slits.nslits):
             # Skip masked
-            if slits.mask[slit_idx] != 0:
+            if (slits.mask[slit_idx] != 0) or (spat_bsplines[slit_idx].breakpoints is None):
```
Comment (jhennawi) [inline pypeit/flatfield.py]: Shouldn't the slit be already masked if (spat_bsplines[slit_idx].breakpoints is None)?
So why is slits.mask[slit_idx] set to True? It seems there is some flaw in the masking logic, but perhaps I'm missing something.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Please see my comment.

**Review by profxj — APPROVED** [review]
approving as I'm sure we'll merge once
the comment from @jhennawi is resolved.

File: pypeit/flatfield.py (hunk 1 above)
Comment (debora-pe) [reply pypeit/flatfield.py]: @jhennawi 
Yes, I think there is something missing in the masking process.
It is not totally clear to me, but what I understood is that, when the master frames are reloaded the bitmask is re-initialized:
`self.slits.mask = self.slits.mask_init.copy()`   #from calibration.py line 612.
Then, Pypeit goes over all the BPM flags saved in the Master frames and re-apply the bitmask. At this point the masking due to the missing bspline fits is not found, and therefore not applied, and `slits.mask[slit_idx]` is then `True` for the slit with missing bspline.

Comment (jhennawi) [reply pypeit/flatfield.py]: Okay, well I'd rather we fix this issue at the source then introduce this line here which at best seems like a kludge. It seems like you are pretty close, since it has to do with the reapplication of the bit mask. There is a clearly a conceptual bug somewhere in the slits.mask or slits.mask_init

Comment (debora-pe) [reply pypeit/flatfield.py]: @jhennawi @profxj 
I understood a little bit better what the problem is. It happens only for morfire (or IR reduction).

The problem is when the masterflat is created. The flatfield code skips masked slits (i.e., slits with ``self.slits.mask == True``).
So in the case of a mosfire reduction, if the first science frame has a slit with failed wavelength calibration (which is computed from the science frame), the flatfield fit is skipped for that slit. But then when the reduction continues to another science frame for which the same slit has a successful wavelength calibration, the masterflat previously created will not have bspline fit for that slit and the reduction crashes.

1.  One solution could be not to skip the flatfield fit for slits with failed wavelength calibration, so that if another frame has the same slit with a successful wavelength calibration, the flatfield can be applied to that slit.
2. Or, we introduce a flatfield BPM flag for when the flatfiled fit is skipped and, if another frame has the same slit with a successful wavelength calibration, the flatfiled will still not be applied.

I think the first one could be a good solution. What do you think?

Comment (jhennawi) [reply pypeit/flatfield.py]: @deborape this is not specific to MOSFIRE. It is rather specific to any situation for which we have multiple arc frames associated with science files as opposed to just one for every single file (i.e. the 'all' option). For mosfire the science is the arc, but this is also sometimes the case for non-IR instruments. Also if a user takes arcs on the sky and wants to associate them with science frames via calibration IDs, this same bug could arise. 

My understanding is that wv_calib is not currently required for flat-fielding, so I'm not sure why the flat is masked for this slit. So in that sense you proposal for 1. seems to make sense. However, we may decide to make wv_calib a dependency if we decide to do something about the vignetting issue that @marlageha and @profxj were discussing. This is not in the code yet, so I'm fine with 1).

Comment (debora-pe) [reply pypeit/flatfield.py]: @jhennawi  @profxj  

I have decided to go with the option 2 and give a flat field-specific BPM flag to the slit for which the flat field is not performed because of bad wave or tilts calibration. In this way the masking will be propagated to every frames.
The bad thing about this is that if there is only one frame for which one slit has bad wave or tilts calibration all the other frames will have that slit masked. However, it make sense if these frames are combined or subtracted with each other.

I think this is the only option at the moment since the tilts image is used during the flat fielding. I made the change and tested on the dataset of the PypeIt user and resolve the issue.
I'll run the tests.

**Comment by debora-pe** [comment]
All good with the Dev Suite 
```
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-09-10T07:47:57.157738
Testing Completed at 2021-09-10T11:47:32.182572
Total Time: 3:59:35.024834
```
but one non related unit tests fails
```
pytest . -W ignore
=============================== test session starts ================================
platform darwin -- Python 3.8.5, pytest-6.1.2, py-1.9.0, pluggy-0.13.1
rootdir: /Users/dpelliccia/PypeIt/PypeIt, configfile: setup.cfg
collected 288 items                                                                

test_alignments.py .                                                         [  0%]
test_arc.py .                                                                [  0%]
test_archive.py ..                                                           [  1%]
test_arcimage.py ..                                                          [  2%]
test_biasframe.py ..                                                         [  2%]
test_bitmask.py ....                                                         [  4%]
test_bpmimage.py ..                                                          [  4%]
test_bspline.py ..........                                                   [  8%]
test_buildimage.py .                                                         [  8%]
test_calibrations.py .......                                                 [ 11%]
test_chk_calibs.py ..                                                        [ 11%]
test_coadd.py .                                                              [ 12%]
test_collate_1d.py ............                                              [ 16%]
test_cooked.py .                                                             [ 16%]
test_coremeta.py .                                                           [ 17%]
test_datacontainer.py ........                                               [ 19%]
test_detector.py ...                                                         [ 20%]
test_display.py .                                                            [ 21%]
test_edgetrace.py .                                                          [ 21%]
test_fitting.py ....                                                         [ 22%]
test_flatfield.py .                                                          [ 23%]
test_flexure.py ..                                                           [ 23%]
test_flux.py ....                                                            [ 25%]
test_fluxspec.py .....                                                       [ 27%]
test_frametype.py .                                                          [ 27%]
test_history.py ....                                                         [ 28%]
test_load.py .                                                               [ 29%]
test_load_images.py ................                                         [ 34%]
test_maskdesign_matching.py ...                                              [ 35%]
test_masterframe.py .                                                        [ 36%]
test_match.py ....                                                           [ 37%]
test_metadata.py ......                                                      [ 39%]
test_moment.py ......                                                        [ 41%]
test_msgs.py ..                                                              [ 42%]
test_opticalmodel.py ....                                                    [ 43%]
test_parse.py ...                                                            [ 44%]
test_pca.py ........                                                         [ 47%]
test_processimage.py ..                                                      [ 48%]
test_procimg.py .                                                            [ 48%]
test_pydl.py .                                                               [ 48%]
test_pypeit.py ..                                                            [ 49%]
test_pypeitimage.py .                                                        [ 50%]
test_pypeitpar.py ................................                           [ 61%]
test_pypeitsetup.py .......F                                                 [ 63%]
test_qa.py .                                                                 [ 64%]
test_runpypeit.py ..                                                         [ 64%]
test_scienceimage.py ...                                                     [ 65%]
test_scripts.py ..............                                               [ 70%]
test_setups.py .....................                                         [ 78%]
test_skysub.py ..                                                            [ 78%]
test_slitmask.py ...                                                         [ 79%]
test_slittrace.py ....                                                       [ 81%]
test_spec2dobj.py ......                                                     [ 83%]
test_specobj.py .......                                                      [ 85%]
test_specobjs.py .....                                                       [ 87%]
test_spectrographs.py ....................                                   [ 94%]
test_syncspec.py .                                                           [ 94%]
test_telluric.py ..                                                          [ 95%]
test_traceimage.py .                                                         [ 95%]
test_utils.py ...                                                            [ 96%]
test_vcorr.py ..                                                             [ 97%]
test_wavemodel.py .                                                          [ 97%]
test_wavetilts.py ....                                                       [ 99%]
test_wvcalib.py ..                                                           [100%]

===================================== FAILURES =====================================
_____________________________ test_run_on_bad_headers ______________________________

    @dev_suite_required
    def test_run_on_bad_headers():
        files = get_lrisr_files()
        # Init
        setupc = pypeitsetup.PypeItSetup(files,
                                         spectrograph_name='keck_lris_red',
                                         path=data_path(''))
        # Run
        par, spectrograph, fitstbl = setupc.run(setup_only=True,
                                                sort_dir=data_path(''))
        # Test
        assert par is None
>       assert setupc.fitstbl['ra'][-1] is None
E       assert 161.17945833333332 is None

/Users/dpelliccia/PypeIt/PypeIt/pypeit/tests/test_pypeitsetup.py:166: AssertionError
------------------------------- Captured stderr call -------------------------------
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.05529.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.05589.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.05649.fits.gz
[WARNING] :: spectrograph.py 823 get_meta_value() - Bad Header, but we'll try to continue on..
[WARNING] :: spectrograph.py 823 get_meta_value() - Bad Header, but we'll try to continue on..
[WARNING] :: spectrograph.py 823 get_meta_value() - Bad Header, but we'll try to continue on..
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.05709.fits.gz
[WARNING] :: spectrograph.py 823 get_meta_value() - Bad Header, but we'll try to continue on..
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07348.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07412.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07470.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07529.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07587.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07646.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07703.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07762.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07820.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07878.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.07937.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.13991.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.14090.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.14167.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.14244.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.14322.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.14399.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.17613.fits.gz
[INFO]    :: metadata.py 203 _build() - Added metadata for LR.20160216.40478.fits.gz
[INFO]    :: metadata.py 1169 get_frame_types() - Typing files
[WARNING] :: keck_lris.py 247 check_frame_type() - Cannot determine if frames are of type align.
[WARNING] :: keck_lris.py 247 check_frame_type() - Cannot determine if frames are of type sky.
[INFO]    :: metadata.py 1225 get_frame_types() - Couldn't identify the following files:
[INFO]    :: metadata.py 1227 get_frame_types() - LR.20160216.07348.fits.gz
[INFO]    :: metadata.py 1232 get_frame_types() - Typing completed!
[INFO]    :: metadata.py 627 unique_configurations() - Using metadata to determine unique configurations.
[INFO]    :: metadata.py 637 unique_configurations() - Unique configurations ignore frames with type: ['bias', 'dark']
[INFO]    :: metadata.py 682 unique_configurations() - Found 1 unique configurations.
[INFO]    :: pypeitsetup.py 467 write() - Wrote sorted file data to /Users/dpelliccia/PypeIt/PypeIt/pypeit/tests/files/keck_lris_red.sorted
[INFO]    :: pypeitsetup.py 493 write() - Setup is complete.
[INFO]    :: pypeitsetup.py 494 write() - Inspect the .sorted file
============================= short test summary info ==============================
FAILED test_pypeitsetup.py::test_run_on_bad_headers - assert 161.17945833333332 i...
==================== 1 failed, 287 passed in 1126.86s (0:18:46) ====================
```

**Comment by profxj** [comment]
That test is on me.  @jhennawi -- consider approving and merging.

File: pypeit/flatfield.py (hunk 1 above)
Comment (jhennawi) [reply pypeit/flatfield.py]: Hi @deborape can we try to discuss via Slack. Basically I agree with your approach except for the following point: 

The dependency of flat fielding on wave_calib is not necessary and was added only to deal with the a not so common case where the user wants to modify the flat in a wavelength dependent way. There is nothing wrong with this in principle, but you can see how adding this as a required argument rather than an optional argument then leads to other undesirable side-effects. Now if a single wavelength calibration fails for a slit, it will now not be reduced for all silts as you point out. But this is very bad, since in fact wave_calib is not required for flat fielding unless the user wants to specify wavelength dependent tweaks (which is rare). This is a very good illustration of why one should think very carefully before adding dependencies on other pypeit objects as required parameters. 

My suggestion is that: 

1) wv_calib should be made optional in flat fielding. 
2) Your approach should be implemented for tilts, but not for wave_calib unless wave_calib is actually required by flat fielding. This is only if pixelflat_min_wave or pixelflat_max_wave are specified in the flat field parset. 

So basically if tilts fails for a single slit we follow your approach, but not if wavelength calib fails unless wavelength calibration was actually needed for the flat (which in the majority of cases it is not). 

The logic here is a bit subtle, so I'm happy to discuss with you and @profxj.

File: pypeit/flatfield.py [hunk 2]
```diff
@@ -687,8 +687,16 @@ def fit(self, spat_illum_only=False, debug=False):
         # Model each slit independently
         for slit_idx, slit_spat in enumerate(self.slits.spat_id):
             # Is this a good slit??
-            if self.slits.mask[slit_idx] != 0:
+            if self.flatpar['pixelflat_min_wave'] is not None or self.flatpar['pixelflat_max_wave'] is not None:
```
Comment (jhennawi) [inline pypeit/flatfield.py]: This block could be shortened to read: 

good_bpms = [0,2] if ( )  else [0, 2, 8, 10]

File: pypeit/flatfield.py [hunk 3]
```diff
@@ -687,8 +687,16 @@ def fit(self, spat_illum_only=False, debug=False):
...
+                # In this case we consider good 'BOXSLIT' (bpm=2)
+                good_bpms = [0, 2]
+            else:
+                # In this case we consider good 'BOXSLIT' (bpm=2), bad wavecalib (bpm=8)
+                # and the combination of the 2 (bpm=10)
+                good_bpms = [0, 2, 8, 10]
```
Comment (jhennawi) [inline pypeit/flatfield.py]: I'm not quite following the logic here and what these good_bpms mean. Rather than checking 8 and 10, I think you should be using the bit arithmetic routines that @kbwestfall developed to check for the relevant bits. The code would also be clearer and more robust if you operated on the bit names rather than the bit integers, since that mapping could change. Finally, I don't get why a BOXSLIT is now a good_bpm whereas it was not before. We don't flat field the box slits?

File: pypeit/wavetilts.py [hunk 4]
```diff
@@ -215,9 +215,8 @@ def __init__(self, mstilt, slits, spectrograph, par, wavepar, det=1, qa_path=Non
         # TODO -- Discuss further with JFH
         all_left, all_right, mask = self.slits.select_edges(initial=True, flexure=self.spat_flexure)  # Grabs all, initial slits
         # self.tilt_bpm = np.invert(mask == 0)
-        # We want to keep the 'BOXSLIT', for which mask value is 2. But we don't want to keep 'BOXSLIT'
-        # with other bad flag (for which the mask value would be > 2)
-        self.tilt_bpm = mask > 2
+        # we consider good slits 'BOXSLIT' (bpm=2), bad wavecalib (bpm=8) and the combination of the 2 (bpm=10)
+        self.tilt_bpm = (mask != 0) & (mask != 2) & (mask != 8) & (mask != 10)
```
Comment (jhennawi) [inline pypeit/wavetilts.py]: Again, use the bit arithmetic routines and bit names rather than hard wired bits. Same comment as above, i.e. we usually don't the tilts for box slits right? This may be something we added for DEIMOS to register the images (as we discussed before), but even then we need to distinguish between box slits where stars overlap and box slits where they don't which can be reduced.

**Review by jhennawi — CHANGES_REQUESTED** [review]
Another iteration @deborape. Feel free to respond in Slack or we can try to discuss as well.

**Review by jhennawi — APPROVED** [review]
Thanks @deborape this looks great. I appreciate your patience with all of my requests!

**Comment by debora-pe** [comment]
All good with the tests

Dev Suite

```
--- PYPEIT DEVELOPMENT SUITE PASSED 128/128 TESTS (Masters ignored) ---
Testing Started at 2021-09-13T21:17:29.377410
Testing Completed at 2021-09-14T01:18:04.255611
Total Time: 4:00:34.878201
```

And unit tests

`===================== 289 passed in 1060.22s (0:17:40) ======================`

**Comment by debora-pe** [comment]
merging
